buildSrc/src/main/kotlin/publish-conventions.gradle.kts

```
@@ -0,0 +1,42 @@
+plugins {
+    `maven-publish`
+    signing
+}
+
+// empty xxx-javadoc.jar
+val javadocJar = tasks.register<Jar>("javadocJar") {
+    archiveClassifier.set("javadoc")
+}
+
+publishing {
+    repositories {
+        configureMavenPublication(this, project)
+    }
+
+    publications.withType<MavenPublication>().all {
+        pom.configureMavenCentralMetadata(project)
+        signPublicationIfKeyPresent(project, this)
+        // add empty javadocs
+        if (name != "kotlinMultiplatform") { // The root module gets the JVM's javadoc JAR
+            artifact(javadocJar)
+        }
+    }
+
+    tasks.withType<AbstractPublishToMaven>().configureEach {
+        dependsOn(tasks.withType<Sign>())
+    }
+
+    // NOTE: This is a temporary WA, see KT-61313.
+    tasks.withType<Sign>().configureEach sign@{
+        val pubName = name.substringAfter("sign").substringBefore("Publication")
+
+        // Task ':linkDebugTest<platform>' uses this output of task ':sign<platform>Publication' without declaring an explicit or implicit dependency
+        tasks.findByName("linkDebugTest$pubName")?.configure<Task> {
+            mustRunAfter(this@sign)
+        }
+        // Task ':compileTestKotlin<platform>' uses this output of task ':sign<platform>Publication' without declaring an explicit or implicit dependency
+        tasks.findByName("compileTestKotlin$pubName")?.configure<Task> {
+            mustRunAfter(this@sign)
+        }
+    }
```

```suggestion
    // NOTE: This is a temporary WA, see KT-61313.
    tasks.withType<Sign>().configureEach sign@{
        val pubName = name.substringAfter("sign").substringBefore("Publication")

        // Task ':linkDebugTest<platform>' uses this output of task ':sign<platform>Publication' without declaring an explicit or implicit dependency
        tasks.findByName("linkDebugTest$pubName")?.configure<Task> {
            mustRunAfter(this@sign)
        }
        // Task ':compileTestKotlin<platform>' uses this output of task ':sign<platform>Publication' without declaring an explicit or implicit dependency
        tasks.findByName("compileTestKotlin$pubName")?.configure<Task> {
            mustRunAfter(this@sign)
        }
    }
```

I think that the first workaround isn't necessary, the second one should work.

---

Actually I'm not sure that we don't need this:
```
 tasks.withType<AbstractPublishToMaven>().configureEach {
        dependsOn(tasks.withType<Sign>())
    }
 ```
 because the second workaround adds `dependsOn` only for mpp test tasks, but we add this plugin for all maven publications.

---

They're both workarounds for this issue https://github.com/gradle/gradle/issues/26132

The error `Some problems were found with the configuration of task [...] Task ':linkDebugTestLinuxX64' uses this output of task ':signLinuxX64Publication' without declaring an explicit or implicit dependency` only really occurs in Kotlin Multiplatform projects. 

This workaround will work, but it isn't optimal:

```kotlin
tasks.withType<AbstractPublishToMaven>().configureEach {
    dependsOn(tasks.withType<Sign>())
}
```

It causes _all_ signing tasks to run, even when if they might not be necessary.

This workaround is better:

```kotlin
tasks.withType<AbstractPublishToMaven>().configureEach {
    mustRunAfter(tasks.withType<Sign>())
}
```

Now the signing tasks will only run if required (as normal), but there's no chance of overlapping outputs because Gradle will make sure _all_ signing tasks run before _any_ publishing tasks.

The other workaround is even more optimised, and will only add the `mustRunAfter()` fix to the specifically affect KMP tasks.

I think that the `mustRunAfter(tasks.withType<Sign>())` workaround is a good compromise, and we should choose it. It's simple, and just works, without having to do string-based checks (which can be flakey). It's not as optimised, but it's not a big deal.

---

I get it, there was no:
```
tasks.withType<AbstractPublishToMaven>().configureEach {
    mustRunAfter(tasks.withType<Sign>())
}
```
in suggested changes, so I thought you just removed it

---

I've applied:
```
tasks.withType<AbstractPublishToMaven>().configureEach {
    mustRunAfter(tasks.withType<Sign>())
}
```

---

https://github.com/Kotlin/kotlinx-atomicfu/pull/435/files#r1610189771